<div style="border-radius:10px; border:#33a0ff solid; padding: 15px; background-color: #f0f1ff; font-size:100%; text-align:left">

<h1> Used Car Listing Price Prediction</h1>
    
</div>

<center><img src="https://raw.githubusercontent.com/anthonynamnam/anthonynamnam/main/icons/image/car-banner.png" alt="memes" width="600" /></center>

---

<div style="border-radius:10px; border:#33a0ff solid; padding: 15px; background-color: #f0f1ff; font-size:100%; text-align:left">

<h2> Project Overview</h2>
    
Please kindly refer to the github repo of this project: <a href="https://github.com/anthonynamnam/brainstation_capstone#project-overview">Link</a>

    
</div>


---

<div style="border-radius:10px; border:#33a0ff solid; padding: 15px; background-color: #f0f1ff; font-size:100%; text-align:left">

<h2> Notebook Overview</h2>
      
In this notebook, we will guide you through the following steps of working with data:

<ol>
    <font size=3><li><b>Feature Engineering 🔨</b></li></font>
    <p>Feature engineering is the catalyst for enhanced predictive power, allowing us to build more accurate models and extract richer insights from our data. Through techniques like dimensionality reduction, one-hot encoding, and creating interaction terms, we will transform our data into a refined masterpiece ready for the final stages of modeling and interpretation.</p>
    <font size=3><li><b>Data Splitting ✂️</b></li></font>
    <p>Data splitting is where we set the stage for the grand performance of our models. Just as a play has acts and scenes, our data is divided into training, validation, and testing sets, each with its unique role. Through strategic data splitting, we ensure our models are not only well-prepared but also ready to shine in the spotlight of real-world scenarios. This section will guide you through the process of splitting data for effective model training and evaluation.</p>

</ol>

So, let's embark on this creative journey, where we'll craft features that will not only optimize our algorithms but also deepen our understanding of the underlying phenomena captured within the data.

    
    
</div>

---

<div style="border-radius:10px; border:#33a0ff solid; padding: 15px; background-color: #f0f1ff; font-size:100%; text-align:left">

<a class="anchor" id="3-toc">
    <h2> Table of Contents </h2>
</a>
    
<ul>    
    <li> <a href="#3-setup">Notebook Set Up</a></li>
    <li> <a href="#3-func">Functions</a></li>
    <li> <a href="#3-load">Data Loading</a></li>
    <li> <a href="#3-feat-transform">Feature Transformation</a></li>
    <li> <a href="#3-split-data">Split Data</a></li>
    <li> <a href="#3-learn">Learning/Takeaway</a></li>
</ul>
    
</div>

---

<div style="border-radius:10px; border:#33a0ff solid; padding: 15px; background-color: #f0f1ff; font-size:100%; text-align:left">
<a class="anchor" id="3-setup">
    <h2> Set Up </h2>
</a>
<b>Table of Content:</b>
<ul>    
    <li> <a href="#3-import">Import Library</a></li>
    <li> <a href="#3-const">Global Const</a></li>
</ul>
</div>

<div style="border-radius:10px; border:#33a0ff solid; padding: 15px; background-color: #f0f1ff; font-size:100%; text-align:left">
<a class="anchor" id="3-import">
<h3> Import Library </h3>
</a>
</div>

In [309]:
import random

# Data Science Package
import numpy as np
import pandas as pd

import warnings

<div style="border-radius:10px; border:#33a0ff solid; padding: 15px; background-color: #f0f1ff; font-size:100%; text-align:left">
<a class="anchor" id="3-const">
<h3> Global Constant </h3>
</a>
</div>

In [310]:

pd.options.display.max_columns = None
ran = random.Random()
ran.seed(42)
warnings.filterwarnings('ignore')

[Back-to-top](#3-toc)

---

<div style="border-radius:10px; border:#33a0ff solid; padding: 15px; background-color: #f0f1ff; font-size:100%; text-align:left">
<a class="anchor" id="3-func">
    <h2> Functions </h2>
</a>
<b>Table of Content:</b>
<ul>    
    <li> <a href="#3-func-print">Helper Funcntions (Print Info)</a></li>
    <li> <a href="#3-func-edit">Helper Funcntions (Edit Dataframe)</a></li>
</ul>
</div>

<div style="border-radius:10px; border:#33a0ff solid; padding: 15px; background-color: #f0f1ff; font-size:100%; text-align:left">
<a class="anchor" id="3-func-print">
<h3> Helper Funcntions (Print Info) </h3>
</a>
</div>

In [311]:
# Helper Functions to print df info and statement
import pandas as pd

def print_num_row(df: pd.DataFrame) -> None:
    """
    Description
    -----
    Retrieve the number of rows of dataframe and print it as a statement.
    
    Args
    -----
    df (pd.DataFrame): the target dataframe
    
    Returns
    -----
    None
    
    Example
    -----
    df = pd.DataFrame(data = {"height":[147,190],"weight":[47,72],"age":[12,28]},index = [0,1])
    print_num_row(df)  =>
        |
        | "The dataframe has 2 rows of record now."
        |
    
    
    """
    print(f"The dataframe has {df.shape[0]} rows of record now.")
    return
    

def print_num_col(df: pd.DataFrame) -> None:
    """
    Description
    -----
    Retrieve the number of columns of dataframe and print it as a statement.
    
    Args
    -----
    df (pd.DataFrame): the target dataframe
    
    Returns
    -----
    None
    
    Example
    -----
    df = pd.DataFrame(data = {"height":[147,190],"weight":[47,72],"age":[12,28]},index = [0,1])
    print_num_col(df) => 
        |
        | "The dataframe has 3 columns now."
        |
    
    
    """
    print(f"The dataframe has {df.shape[1]} columns now.")
    return
        
def print_dim(df: pd.DataFrame) -> None:
    """
    Description
    -----
    Retrieve the shape of dataframe and print it as a statement.
    
    Args
    -----
    df (pd.DataFrame): the target dataframe
    
    Returns
    -----
    None
    
    Example
    -----
    abc_df = pd.DataFrame(data = {"height":[147,190],"weight":[47,72],"age":[12,28]},index = [0,1])
    print_dim(abc_df) =>
        |
        | "There are 2 rows and 3 columns in this dataframe now."
        |
    
    
    """
    print(f"There are {df.shape[0]} rows and {df.shape[1]} columns in this dataframe now.")
    return


def print_null_count(df: pd.DataFrame,cols:list = []) -> None:
    """
    Description
    -----
    Count the null value in each columns.
    If `cols` is provided, only show the null value count for the columns in `cols`. 
    Otherwise, show null value count for all columns.
    
    Args
    -----
    df (pd.DataFrame): target dataframe
    cols (list): the column names to show the null value count. Default: []
    
    Returns
    -----
    None
    
    Example
    -----
    abc_df = pd.DataFrame(data = {"height":[147,190],"weight":[47,np.nan],"age":[np.nan,28]},index = [0,1])
    print_null_count(abc_df) => 
        |
        | === Null Count ===
        | height    0
        | weight    1
        | age       2
        | dtype: int64
        |
        
    print_null_count(abc_df,cols=["age"]) => 
        |
        | === Null Count ===
        | Column `age`: 2
        |
        
    print_null_count(abc_df,cols=["age","weight"]) => 
        |
        | === Null Count ===
        | Column `age`: 2
        | Column `weight`: 1
        |
    
    """
    if len(cols) == 0:
        null_count = df.isnull().sum()
        
        print("=== Null Count ===")
        print(null_count)
    else:
        assert set(cols).issubset(df.columns)
        null_count = df[cols].isnull().sum()
        
        print("=== Null Count ===")
        for col in cols:
            print(f"Column `{col}`: {null_count[col]}")
    return 


def print_null_pct(df: pd.DataFrame,cols:list = []) -> None:
    """
    Description
    -----
    Count the null percentage in each columns.
    If `cols` is provided, only show the null percentage for the columns in `cols`. 
    Otherwise, show null percentage for all columns.
    
    Args
    -----
    df (pd.DataFrame): target dataframe
    cols (list): the column names to show the null percentage. Default: []
    
    Returns
    -----
    None
    
    Example
    -----
    abc_df = pd.DataFrame(data = {"height":[147,190],"weight":[47,np.nan],"age":[np.nan,np.nan]},index = [0,1])
    print_null_pct(abc_df) => 
        |
        | === Null Count Precentage ===
        | height      0.0%
        | weight     50.0%
        | age       100.0%
        | dtype: object
        |
        
    print_null_pct(abc_df,cols=["weight"]) => 
        |
        | === Null Count Precentage ===
        | Column weight: 50.0%
        |
    
    """
    total_row = df.shape[0]
    if len(cols) == 0:
        null_count = df.isnull().sum()
        null_pct = null_count / total_row * 100
        
        print("=== Null Count Precentage ===")
        print(null_pct.round(2).astype(str)+"%")
    else:
        assert set(cols).issubset(df.columns)
        null_count = df[cols].isnull().sum()
        null_pct = null_count / total_row * 100

        print("=== Null Count Precentage ===")
        for col in cols:
            print(f"Column {col}: {round(null_pct[col],4)}%")
    return

def print_duplicated_count(df: pd.DataFrame) -> None:
    """
    Description
    -----
    Count the number of duplicated rows.
    
    Args
    -----
    df (pd.DataFrame): target dataframe
    
    Returns
    -----
    None
    
    Example
    -----
    abc_df = pd.DataFrame(data = {"height":[147,190,147],"weight":[47,np.nan,47],"age":[13,27,13]},index = [0,1,2])
    print_duplicated_count(abc_df) => 
        |
        | There are 1 duplicated rows
        | 
        
    """
    print(f"There are {df.duplicated().sum()} duplicated rows")
    return



<div style="border-radius:10px; border:#33a0ff solid; padding: 15px; background-color: #f0f1ff; font-size:100%; text-align:left">
<a class="anchor" id="3-func-edit">
<h3> Helper Funcntions (Edit Dataframe) </h3>
</a>
</div>

In [312]:
def drop_cols_if_exist(df: pd.DataFrame,cols_to_drop:list) -> pd.DataFrame:
    """
    Description
    -----
    Drop a column from a dataframe with inplace = True. Only execute the dropping if the cols exist.
    
    Args
    -----
    df (pd.DataFrame): the target dataframe
    cols_to_drop (list): the list of column to be dropped
    
    Returns
    -----
    df (pd.DataFrame): the dataframe with columns dropped
    
    Example
    -----
    # Create a DataFrame
    abc_df = pd.DataFrame(data = {"height":[147,190,147],"weight":[47,np.nan,47],"age":[13,27,13]},index = [0,1,2])
    print(abc_df)  =>
        |
        |    height  weight  age
        | 0     147    47.0   13
        | 1     190     NaN   27
        | 2     147    47.0   13
        |
        
    # Drop columns if exist
    dropped_abc_df = drop_cols_if_exist(abc_df,cols_to_drop=["weight"])
    print(dropped_abc_df)   =>
        | Successfully dropped columns: {'weight'}
        |    height  age
        | 0     147   13
        | 1     190   27
        | 2     147   13
    
    
    """
    intersect_cols = set(cols_to_drop).intersection(df.columns)
    df.drop(columns=intersect_cols,inplace=True)
    print(f"Successfully dropped columns: {intersect_cols}")
    return df    

[Back-to-top](#3-toc)

---

<div style="border-radius:10px; border:#33a0ff solid; padding: 15px; background-color: #f0f1ff; font-size:100%; text-align:left">
<a class="anchor" id="3-load">
    <h2> Data Loading </h2>
</a>
<b>Table of Content:</b>
<ul>    
    <li> <a href="#3-load-process">Load Processed Data</a></li>
    <li> <a href="#3-san-check-1">Sanity Check</a></li>
</ul>
</div>

<div style="border-radius:10px; border:#33a0ff solid; padding: 15px; background-color: #f0f1ff; font-size:100%; text-align:left">
<a class="anchor" id="3-load-process">
<h3> Load the processed data </h3>
</a>
</div>

In [313]:
# Define dtype for cleaned dataset
PROCESSED_COL_DTYPE = {
    "vin":str,
    "miles":float,
    "year":int,
    "make":str,
    "model":str,
    "trim":str,
    "body_type":str,
    "vehicle_type":str,
    "drivetrain":str,
    "transmission":str,
    "engine_size":float,
    "engine_block":str,
    "city":str,
    "state":str,
    "price_range":int,
    
    # For encoded fuel type
    "fuel_M85":int,
    "fuel_Lpg":int,
    "fuel_Diesel":int,
    "fuel_Unleaded":int,
    "fuel_Hydrogen":int,
    "fuel_PremiumUnleaded":int,
    "fuel_Biodiesel":int,
    "fuel_E85":int,
    "fuel_Electric":int,
    "fuel_CompressedNaturalGas":int,
}

In [314]:
# First, we read the cleaned dataset
df = pd.read_parquet(path = "data/processed-data.parquet",
                     columns = PROCESSED_COL_DTYPE)
df.reset_index(drop = True, inplace=True)

<div style="border-radius:10px; border:#33a0ff solid; padding: 15px; background-color: #f0f1ff; font-size:100%; text-align:left">
<a class="anchor" id="3-san-check-1">
<h3> Sanity Check </h3>
</a>
</div>

In [315]:
df.head()

vin    miles    year   make model  trim body_type  \
0  19UNC1B01HY800062   9966.0  2017.0  Acura   NSX  Base     Coupe   
1  19UNC1B02HY800023   5988.0  2017.0  Acura   NSX  Base     Coupe   
2  19UNC1B02HY800071  24242.0  2017.0  Acura   NSX  Base     Coupe   
3  19UNC1B02LY800001   6637.0  2020.0  Acura   NSX  Base     Coupe   
4  19UNC1B03HY800063  18281.0  2017.0  Acura   NSX  Base     Coupe   

  vehicle_type drivetrain transmission  engine_size engine_block  \
0          Car        4WD    Automatic          3.5            V   
1          Car        4WD    Automatic          3.5            V   
2          Car        4WD    Automatic          3.5            V   
3          Car        4WD    Automatic          3.5            V   
4          Car        4WD    Automatic          3.5            V   

                  city state  price_range  fuel_M85  fuel_Lpg  fuel_Diesel  \
0           Edmundston    NB            5         0         0            0   
1  Notre-dame-des-pins    QC            5         0         0            0   
2            Coquitlam    BC            5         0         0            0   
3            Pickering    ON            5         0         0            0   
4              Montral    QC            5         0         0            0   

   fuel_Unleaded  fuel_Hydrogen  fuel_PremiumUnleaded  fuel_Biodiesel  \
0              0              0                     1               0   
1              0              0                     1               0   
2              0              0                     1               0   
3              0              0                     1               0   
4              0              0                     1               0   

   fuel_E85  fuel_Electric  fuel_CompressedNaturalGas  
0         0              1                          0  
1         0              1                          0  
2         0              1                          0  
3         0              1                          0  
4         0              1                          0

In [316]:
print_dim(df)

There are 4141534 rows and 25 columns in this dataframe now.


In [317]:
print_null_count(df)

=== Null Count ===
vin                          0
miles                        0
year                         0
make                         0
model                        0
trim                         0
body_type                    0
vehicle_type                 0
drivetrain                   0
transmission                 0
engine_size                  0
engine_block                 0
city                         0
state                        0
price_range                  0
fuel_M85                     0
fuel_Lpg                     0
fuel_Diesel                  0
fuel_Unleaded                0
fuel_Hydrogen                0
fuel_PremiumUnleaded         0
fuel_Biodiesel               0
fuel_E85                     0
fuel_Electric                0
fuel_CompressedNaturalGas    0
dtype: int64


In [318]:
print_duplicated_count(df)

There are 0 duplicated rows


[Back-to-top](#3-toc)

---

<div style="border-radius:10px; border:#33a0ff solid; padding: 15px; background-color: #f0f1ff; font-size:100%; text-align:left">
<a class="anchor" id="3-feat-transform">
    <h2> Features Transformation </h2>
</a>
<b>Table of Content:</b>
<ul>    
    <li> <a href="#3-log-miles">Log Transformation - <b><i>miles</i></b></a></li>
    <li> <a href="#3-vehicle-history">Same Vehicle History Transformation</a></li>
    <li> <a href="#3-drop-vin">Remove Column - <b><i>vin</i></b></a></li>
</ul>
</div>

<div style="border-radius:10px; border:#33a0ff solid; padding: 15px; background-color: #f0f1ff; font-size:100%; text-align:left">
<a class="anchor" id="3-log-miles">
    <h3> Log Transformation on <b><i>miles</i></b> </h3>
</a>
</div>

In [319]:
df.head()

vin    miles    year   make model  trim body_type  \
0  19UNC1B01HY800062   9966.0  2017.0  Acura   NSX  Base     Coupe   
1  19UNC1B02HY800023   5988.0  2017.0  Acura   NSX  Base     Coupe   
2  19UNC1B02HY800071  24242.0  2017.0  Acura   NSX  Base     Coupe   
3  19UNC1B02LY800001   6637.0  2020.0  Acura   NSX  Base     Coupe   
4  19UNC1B03HY800063  18281.0  2017.0  Acura   NSX  Base     Coupe   

  vehicle_type drivetrain transmission  engine_size engine_block  \
0          Car        4WD    Automatic          3.5            V   
1          Car        4WD    Automatic          3.5            V   
2          Car        4WD    Automatic          3.5            V   
3          Car        4WD    Automatic          3.5            V   
4          Car        4WD    Automatic          3.5            V   

                  city state  price_range  fuel_M85  fuel_Lpg  fuel_Diesel  \
0           Edmundston    NB            5         0         0            0   
1  Notre-dame-des-pins    QC            5         0         0            0   
2            Coquitlam    BC            5         0         0            0   
3            Pickering    ON            5         0         0            0   
4              Montral    QC            5         0         0            0   

   fuel_Unleaded  fuel_Hydrogen  fuel_PremiumUnleaded  fuel_Biodiesel  \
0              0              0                     1               0   
1              0              0                     1               0   
2              0              0                     1               0   
3              0              0                     1               0   
4              0              0                     1               0   

   fuel_E85  fuel_Electric  fuel_CompressedNaturalGas  
0         0              1                          0  
1         0              1                          0  
2         0              1                          0  
3         0              1                          0  
4         0              1                          0

In [320]:
# Add 1 to the column to avoid log(0) = -inf
df["log_miles"] = np.log(df["miles"]+1)

In [321]:
# Drop the miles column
drop_cols_if_exist(df,["miles"])

Successfully dropped columns: {'miles'}


vin    year   make             model        trim  \
0        19UNC1B01HY800062  2017.0  Acura               NSX        Base   
1        19UNC1B02HY800023  2017.0  Acura               NSX        Base   
2        19UNC1B02HY800071  2017.0  Acura               NSX        Base   
3        19UNC1B02LY800001  2020.0  Acura               NSX        Base   
4        19UNC1B03HY800063  2017.0  Acura               NSX        Base   
...                    ...     ...    ...               ...         ...   
4141529  1FT7W2BT8BEA10040  2011.0   Ford  F-250 Super Duty      Lariat   
4141530  1FT7W2BT9KEE35377  2019.0   Ford  F-250 Super Duty         XLT   
4141531  1FD7W2BT0KEC89883  2019.0   Ford  F-250 Super Duty          XL   
4141532  1FT7W2BT5EEB84944  2014.0   Ford  F-250 Super Duty      Lariat   
4141533  1FT7W2BT0BEB86340  2011.0   Ford  F-250 Super Duty  King Ranch   

        body_type vehicle_type drivetrain transmission  engine_size  \
0           Coupe          Car        4WD    Automatic          3.5   
1           Coupe          Car        4WD    Automatic          3.5   
2           Coupe          Car        4WD    Automatic          3.5   
3           Coupe          Car        4WD    Automatic          3.5   
4           Coupe          Car        4WD    Automatic          3.5   
...           ...          ...        ...          ...          ...   
4141529    Pickup        Truck        4WD    Automatic          6.7   
4141530    Pickup        Truck        4WD    Automatic          6.7   
4141531    Pickup        Truck        4WD    Automatic          6.7   
4141532    Pickup        Truck        4WD    Automatic          6.7   
4141533    Pickup        Truck        4WD    Automatic          6.7   

        engine_block                 city state  price_range  fuel_M85  \
0                  V           Edmundston    NB            5         0   
1                  V  Notre-dame-des-pins    QC            5         0   
2                  V            Coquitlam    BC            5         0   
3                  V            Pickering    ON            5         0   
4                  V              Montral    QC            5         0   
...              ...                  ...   ...          ...       ...   
4141529            V           White Pine    TN            2         0   
4141530            V               Irving    TX            4         0   
4141531            V              Ephrata    WA            4         0   
4141532            V                 Pana    IL            4         0   
4141533            V                Tampa    FL            3         0   

         fuel_Lpg  fuel_Diesel  fuel_Unleaded  fuel_Hydrogen  \
0               0            0              0              0   
1               0            0              0              0   
2               0            0              0              0   
3               0            0              0              0   
4               0            0              0              0   
...           ...          ...            ...            ...   
4141529         0            1              0              0   
4141530         0            1              0              0   
4141531         0            1              0              0   
4141532         0            1              0              0   
4141533         0            1              0              0   

         fuel_PremiumUnleaded  fuel_Biodiesel  fuel_E85  fuel_Electric  \
0                           1               0         0              1   
1                           1               0         0              1   
2                           1               0         0              1   
3                           1               0         0              1   
4                           1               0         0              1   
...                       ...             ...       ...            ...   
4141529                     0               0         0              0   
4141530            

<div style="border-radius:10px; border:#33a0ff solid; padding: 15px; background-color: #f0f1ff; font-size:100%; text-align:left">
<a class="anchor" id="3-vehicle-history">
    <h3> Same Vehicle History Transformation </h3>
</a>
</div>

<div style="border-radius:10px; border:#33a0ff solid; padding: 15px; background-color: #f0f1ff; font-size:100%; text-align:left">
<a class="anchor" id="">
    <h4> Multiple Listing Vehicles Count </h4>
</a>
    
- Let's check which cars are with multiple listing.
</div>

In [322]:
df["vin"].value_counts()[0:20]

1C6RR6TT7LS102367    64
1C6RR6TT3LS100471    56
WBY7Z4C51JVC34561    55
1C6RR7TT8LS117773    54
1GNSCCE09DR250565    54
WBY7Z4C57JVD95870    53
1GCGSCEN6J1265015    53
5YJSA1H23EFP62202    52
1G1YY25U975129704    52
WBY7Z4C55JVD95415    52
KL4MMBS2XLB097887    52
1G1PC5SBXE7237877    52
1FT7W2BT5KEF52440    51
1GYKNCRS9LZ193694    51
3GNKBFRS3KS683020    51
3LN6L5LU4KR613914    51
KL4MMDSL4LB107883    51
2G1WC5E34G1127360    50
1GCGTCEN8K1283479    50
1GNERNKW9LJ200956    49
Name: vin, dtype: int64

Let's count how many vehicles have multiple listing.

In [323]:
print(f"There are {(df['vin'].value_counts() > 1).sum()} unique vehicles with more than one listing records.")


There are 702930 unique vehicles with more than one listing records.


<div style="border-radius:10px; border:#33a0ff solid; padding: 15px; background-color: #f0f1ff; font-size:100%; text-align:left">
<a class="anchor" id="">
    <h4> Deal with Multiple Listing </h4>
</a>
    
- Let's use one of the examples to analyse what we should do on these multiple listing vehicles.
- Our example: `vin` = `19UUB3F46LA003699`
</div>

In [324]:
df[(df["vin"]=="19UUB3F46LA003699")].T

1184274             1188071  \
vin                         19UUB3F46LA003699   19UUB3F46LA003699   
year                                   2020.0              2020.0   
make                                    Acura               Acura   
model                                     TLX                 TLX   
trim                       Technology Package  Technology Package   
body_type                               Sedan               Sedan   
vehicle_type                              Car                 Car   
drivetrain                                4WD                 4WD   
transmission                        Automatic           Automatic   
engine_size                               3.5                 3.5   
engine_block                                V                   V   
city                               West Nyack            Portland   
state                                      NY                  OR   
price_range                                 1                   3   
fuel_M85                                    0                   0   
fuel_Lpg                                    0                   0   
fuel_Diesel                                 0                   0   
fuel_Unleaded                               0                   0   
fuel_Hydrogen                               0                   0   
fuel_PremiumUnleaded                        1                   1   
fuel_Biodiesel                              0                   0   
fuel_E85                                    0                   0   
fuel_Electric                               0                   0   
fuel_CompressedNaturalGas                   0                   0   
log_miles                            9.575539            9.263881   

                                      3172290  
vin                         19UUB3F46LA003699  
year                                   2020.0  
make                                    Acura  
model                                     TLX  
trim                       Technology Package  
body_type                               Sedan  
vehicle_type                              Car  
drivetrain                                4WD  
transmission                        Automatic  
engine_size                               3.5  
engine_block                                V  
city                                 Hazleton  
state                                      PA  
price_range                                 3  
fuel_M85                                    0  
fuel_Lpg                                    0  
fuel_Diesel                                 0  
fuel_Unleaded                               0  
fuel_Hydrogen                               0  
fuel_PremiumUnleaded                        1  
fuel_Biodiesel                              0  
fuel_E85                                    0  
fuel_Electric                               0  
fuel_CompressedNaturalGas                   0  
log_miles                            9.575539

<div style="border-radius:10px; border: #ffd500 solid; padding: 15px; background-color: #ffffcf; font-size:100%; text-align:left">
<font size="4px" color="#ffa600"><b>💡 Insight:</b></font>

- For the same car with `vin` of `1GCGSCEN6J1265015`, there are multiple listing records. We do not have any temporal columns to indicate the time of listing, but we can deduce the temporal order of the listing. 
- Even for the listing with same `log_miles`, they have multiple record due to different locations, different supported fuel types.
    - Different **locations** means the same car is listed through multiple dealers, 
    - Different **supported fuel types** means either one of the following cases:
        1. The supported fuel types were wrongly enter (i.e. car not supporting Diesel => `Diesel = `]).
        2. The supported fuel types were omitted (i.e. car supporting Diesel and Unleaded => `Diesel = 1` & `Unleaded = 0`).
- For the same car listing on multiple platforms, it should have the same `miles` travelled on the odometer. This means we may deduce if the listing are at the same time by the `miles` (or `log_miles`) columns.
- For some of the records, the `engine_block` may even change, it may be caused by the auto-part replacement by the car owner.
    - If these records are with different info (`engine_block`, `engine_size` & etc.) of the car, but same `vin` number, then we use majority vote to decide which record to keep.
    - For supported fuel types, as it accept multiple values, we will only keep those support fuel types that with > 50% occurence in all the records with same `log_miles`.
- From here, we will drop both `states` and `city` as the same vehicle may be listed on multiple dealers platform from the same seller.

<div>

In [325]:
df.loc[df["vin"]=="19UUB3F46LA003699","log_miles"].unique()

array([9.57553889, 9.26388114])

<div style="border-radius:10px; border: #ffd500 solid; padding: 15px; background-color: #ffffcf; font-size:100%; text-align:left">
<font size="4px" color="#ffa600"><b>⚠️ Note:</b></font>

From the above example, we can see 2 different values of `log_miles`.

Note that when we applied log-transformation, we added 1 to each values to avoid getting `-inf`. After log-transformation, we had 0, which means it is 0 miles before log-transformation (i.e. `log(0+1) = 0`).

We can deduce the temporal order of listing should be:
1. `log_miles`: 9.26388114  (Earlier Listing)
2.  `log_miles`: 9.57553889 (Latest listing)
    
<div>

In [326]:

def process_same_car_history(same_car_records,results):

    log_miles = same_car_records["log_miles"].unique()

    # new_df = pd.DataFrame()
    # Iterate each group with different log_miles values
    for lm in log_miles:
        temp_df = same_car_records[same_car_records["log_miles"]==lm]

        # Check Duplicates
        uni_records = temp_df.drop(columns=["city","state"]).drop_duplicates()

        if len(uni_records) > 1:
            # Process the history if there are different values in the vehicle info / supported fuel types
            col_replace_dict = {}

            for col in uni_records.columns:

                # Value count on each column
                col_val_count = uni_records[col].value_counts()

                # if there are multiple values in same columns, use the one with maximum occurence to replace it.
                # if they have the same occurrence then random select one.
                if len(col_val_count) > 1:

                    mode = uni_records[col].mode()

                    # Special handling for col price_range
                    if col == "price_range":

                        # If multiple price range appears same # of time, take median and 
                        if len(mode) > 1:
                            col_replace_dict[col] = int(np.round(np.median(mode)+0.000001))
                        else:
                            col_replace_dict[col] = mode.values[0]

                    # For other columns
                    else:
                        col_replace_dict[col] = ran.choice(mode.values)

            for col,val in col_replace_dict.items():
                uni_records[col] = val
            uni_records.drop_duplicates(inplace = True)

        # Save the processed record
        # new_df = pd.concat([new_df,uni_records])
        results.append(uni_records)
        
    return

<div style="border-radius:10px; border: #ffd500 solid; padding: 15px; background-color: #ffffcf; font-size:100%; text-align:left">
<font size="4px" color="#ffa600"><b>⚙️ Before Processing (Example)</b></font>
<div>

In [327]:
df[df['vin']=="19UUB3F46LA003699"]

vin    year   make model                trim body_type  \
1184274  19UUB3F46LA003699  2020.0  Acura   TLX  Technology Package     Sedan   
1188071  19UUB3F46LA003699  2020.0  Acura   TLX  Technology Package     Sedan   
3172290  19UUB3F46LA003699  2020.0  Acura   TLX  Technology Package     Sedan   

        vehicle_type drivetrain transmission  engine_size engine_block  \
1184274          Car        4WD    Automatic          3.5            V   
1188071          Car        4WD    Automatic          3.5            V   
3172290          Car        4WD    Automatic          3.5            V   

               city state  price_range  fuel_M85  fuel_Lpg  fuel_Diesel  \
1184274  West Nyack    NY            1         0         0            0   
1188071    Portland    OR            3         0         0            0   
3172290    Hazleton    PA            3         0         0            0   

         fuel_Unleaded  fuel_Hydrogen  fuel_PremiumUnleaded  fuel_Biodiesel  \
1184274              0              0                     1               0   
1188071              0              0                     1               0   
3172290              0              0                     1               0   

         fuel_E85  fuel_Electric  fuel_CompressedNaturalGas  log_miles  
1184274         0              0                          0   9.575539  
1188071         0              0                          0   9.263881  
3172290         0              0                          0   9.575539

<div style="border-radius:10px; border: #ffd500 solid; padding: 15px; background-color: #ffffcf; font-size:100%; text-align:left">
<font size="4px" color="#ffa600"><b>⚙️ After Processing (Example)</b></font>
<div>

In [328]:
res = []
process_same_car_history(df[df["vin"]=="19UUB3F46LA003699"],res)
pd.concat(res)

vin    year   make model                trim body_type  \
1184274  19UUB3F46LA003699  2020.0  Acura   TLX  Technology Package     Sedan   
1188071  19UUB3F46LA003699  2020.0  Acura   TLX  Technology Package     Sedan   

        vehicle_type drivetrain transmission  engine_size engine_block  \
1184274          Car        4WD    Automatic          3.5            V   
1188071          Car        4WD    Automatic          3.5            V   

         price_range  fuel_M85  fuel_Lpg  fuel_Diesel  fuel_Unleaded  \
1184274            2         0         0            0              0   
1188071            3         0         0            0              0   

         fuel_Hydrogen  fuel_PremiumUnleaded  fuel_Biodiesel  fuel_E85  \
1184274              0                     1               0         0   
1188071              0                     1               0         0   

         fuel_Electric  fuel_CompressedNaturalGas  log_miles  
1184274              0                          0   9.575539  
1188071              0                          0   9.263881

<div style="border-radius:10px; border: #ffd500 solid; padding: 15px; background-color: #ffffcf; font-size:100%; text-align:left">
<font size="4px" color="#ffa600"><b>⚙️ Retrieve vin with more than 2 records</b></font>
<div>

In [329]:
vin_count = df["vin"].value_counts()
vin_count = vin_count[vin_count>1]
vin_count.index

Index(['1C6RR6TT7LS102367', '1C6RR6TT3LS100471', 'WBY7Z4C51JVC34561',
       '1C6RR7TT8LS117773', '1GNSCCE09DR250565', 'WBY7Z4C57JVD95870',
       '1GCGSCEN6J1265015', '5YJSA1H23EFP62202', '1G1YY25U975129704',
       'WBY7Z4C55JVD95415',
       ...
       '5UXKT0C57J0V98840', '3FADP4BJ6EM199821', '3MYDLBYV1HY177440',
       '3HGGK5H61JM712628', '1FMCU0GD3HUB79659', '3FA6P0HR6DR273058',
       '3C6UR5CL9KG593566', '3KPC24A61LE105059', '3GKALMEV4KL205130',
       '3GCUKNEC7HG386404'],
      dtype='object', length=702930)

<div style="border-radius:10px; border: #ffd500 solid; padding: 15px; background-color: #ffffcf; font-size:100%; text-align:left">
<font size="4px" color="#ffa600"><b>⚙️ Process our original dataframe (with multi threading for speeding up)</b></font>

- As there are over 700K vin with multiple listing records, we apply multi-threading to speed up the process. 
    
- Please avoid to run the next cell.
    
- We split all 700K vin into batches of size 100. For each batch, we use multi-threading to perform processing. However, it still take more than 60 hours to process all of them. 
    
- To further speed up, we mannually duplicate the code in multiple notebooks and run them indenepently, this is to simulate the distributed processing as each vin are independent.
    
- All the transformed data are saved at: `data/subprocess`.
    
<div>

In [330]:
## !!! This cell will run for hours !!!
## We will skip this step here

import threading
import time

start_ind = 0
interval = 100

end_ind = start_ind+interval
max_count = len(vin_count.index)

max_thread = 10
basic_thread_count = threading.active_count()
total_max_thread = basic_thread_count + max_thread

while start_ind < max_count:
#     print(f"Starting [{start_ind}-{end_ind}] ", end ="")
#     vins = list(vin_count.index)[start_ind:min(end_ind,max_count)]
#     threads = []
#     results = []
#     start = time.time()

#     while len(vins) > 0:
#         if threading.active_count() < total_max_thread:
#             t = threading.Thread(target=process_same_car_history, args=(df[df["vin"]==vins.pop()],results))
#             threads.append(t)
#             t.start()

#     for t in threads:
#         t.join()

#     stop = time.time()
#     pd.concat(results).to_parquet(path=f"data/subprocess/vehicle-history-processed-{start_ind}-{end_ind}.parquet",index = False)
#     print(f"| Exported Successfully... | Elapsed time: {round(stop - start,2)}s")
    
    # Go to next set
    start_ind += interval
    end_ind += interval

<div style="border-radius:10px; border:#33a0ff solid; padding: 15px; background-color: #f0f1ff; font-size:100%; text-align:left">
<a class="anchor" id="3-join-split">
    <h4> Concatenate the transformed data </h4>
</a>
</div>

In [350]:
import os

l = os.listdir("./data/subprocess/")
l = list(filter(lambda f: (f[-8:]== ".parquet") and ("vehicle-history-processed-" in f) ,l))
num_id_l = sorted(list(map(lambda f: int(f.split(".")[0].split("-")[3]),l)))

range_list = []
start_num = num_id_l[0]
prev_num = num_id_l[0]
for num in num_id_l:
    if num - prev_num > 100 or num == num_id_l[-1]:
        range_list.append(f"{start_num}-{prev_num}")
        start_num = num
        prev_num = num
    else:
        prev_num = num

print(f"Number of files: {len(l)}")


Number of files: 7030


In [332]:
# Define dtype for transformed dataset
VEHICLE_HISTORY_COL_DTYPE = {
    "vin":str,
    "log_miles":float,
    "year":int,
    "make":str,
    "model":str,
    "trim":str,
    "body_type":str,
    "vehicle_type":str,
    "drivetrain":str,
    "transmission":str,
    "engine_size":float,
    "engine_block":str,
    "price_range":int,
    
    # For encoded fuel type
    "fuel_M85":int,
    "fuel_Lpg":int,
    "fuel_Diesel":int,
    "fuel_Unleaded":int,
    "fuel_Hydrogen":int,
    "fuel_PremiumUnleaded":int,
    "fuel_Biodiesel":int,
    "fuel_E85":int,
    "fuel_Electric":int,
    "fuel_CompressedNaturalGas":int,
}

In [334]:
# %%timeit -n 1 -r 1
l_df = []
for f in l:
    # First, we read the cleaned dataset
    temp_df = pd.read_parquet(path = f"data/subprocess/{f}",
                              columns = VEHICLE_HISTORY_COL_DTYPE)
    l_df.append(temp_df)
    
transformed_df = pd.concat(l_df)

In [335]:
# Sanity Check
transformed_df.head()

vin  log_miles    year       make     model          trim  \
0  3TMJU4GNXEM156908  11.652522  2014.0     Toyota    Tacoma     PreRunner   
1  WA1BNAFY2L2055193   8.403576  2020.0       Audi        Q5  Premium Plus   
2  5TFJU4GN0CX025106  11.534560  2012.0     Toyota    Tacoma     PreRunner   
3  1GCGTBEN8K1288868   9.716013  2019.0  Chevrolet  Colorado    Work Truck   
4  5TFJU4GN4EX055759  11.318126  2014.0     Toyota    Tacoma     PreRunner   

  body_type vehicle_type drivetrain transmission  engine_size engine_block  \
0    Pickup        Truck        RWD    Automatic          4.0            V   
1       SUV        Truck        4WD    Automatic          2.0            I   
2    Pickup        Truck        RWD    Automatic          4.0            V   
3    Pickup        Truck        4WD    Automatic          3.6            V   
4    Pickup        Truck        RWD    Automatic          4.0            V   

   price_range  fuel_M85  fuel_Lpg  fuel_Diesel  fuel_Unleaded  fuel_Hydrogen  \
0            2         0         0            0              1              0   
1            3         0         0            0              0              0   
2            2         0         0            0              1              0   
3            3         0         0            0              1              0   
4            2         0         0            0              1              0   

   fuel_PremiumUnleaded  fuel_Biodiesel  fuel_E85  fuel_Electric  \
0                     0               0         0              0   
1                     1               0         0              0   
2                     0               0         0              0   
3                     0               0         0              0   
4                     0               0         0              0   

   fuel_CompressedNaturalGas  
0                          0  
1                          0  
2                          0  
3                          0  
4                          0

In [336]:
# Sanity Check
transformed_df.shape

(757342, 23)

<div style="border-radius:10px; border:#33a0ff solid; padding: 15px; background-color: #f0f1ff; font-size:100%; text-align:left">
<a class="anchor" id="3-drop-col">
    <h4> Retrieve vin with no multiple listing - <b><i>vin</i></b> </h4>
</a>
</div>

In [337]:
# Retrieve vin data with single listing
single_listing_df = df[~df["vin"].isin(vin_count.index)]

# Remove city and state
single_listing_df.drop(columns=["city","state"],inplace = True)

/var/folders/zh/ycxy1lsx5sx0g_2n4l8hfvph0000gn/T/ipykernel_14634/3731808552.py:5: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  single_listing_df.drop(columns=["city","state"],inplace = True)


In [338]:
# Sanity Check
single_listing_df.head()

vin    year   make model  trim body_type vehicle_type  \
0  19UNC1B01HY800062  2017.0  Acura   NSX  Base     Coupe          Car   
1  19UNC1B02HY800023  2017.0  Acura   NSX  Base     Coupe          Car   
2  19UNC1B02HY800071  2017.0  Acura   NSX  Base     Coupe          Car   
3  19UNC1B02LY800001  2020.0  Acura   NSX  Base     Coupe          Car   
4  19UNC1B03HY800063  2017.0  Acura   NSX  Base     Coupe          Car   

  drivetrain transmission  engine_size engine_block  price_range  fuel_M85  \
0        4WD    Automatic          3.5            V            5         0   
1        4WD    Automatic          3.5            V            5         0   
2        4WD    Automatic          3.5            V            5         0   
3        4WD    Automatic          3.5            V            5         0   
4        4WD    Automatic          3.5            V            5         0   

   fuel_Lpg  fuel_Diesel  fuel_Unleaded  fuel_Hydrogen  fuel_PremiumUnleaded  \
0         0            0              0              0                     1   
1         0            0              0              0                     1   
2         0            0              0              0                     1   
3         0            0              0              0                     1   
4         0            0              0              0                     1   

   fuel_Biodiesel  fuel_E85  fuel_Electric  fuel_CompressedNaturalGas  \
0               0         0              1                          0   
1               0         0              1                          0   
2               0         0              1                          0   
3               0         0              1                          0   
4               0         0              1                          0   

   log_miles  
0   9.207035  
1   8.697680  
2  10.095883  
3   8.800566  
4   9.813672

In [339]:
# Sanity Check
single_listing_df.shape

(1595259, 23)

---

In [340]:
joined_df = pd.concat([single_listing_df,transformed_df])

In [341]:
# Sanity Check
joined_df.head()

vin    year   make model  trim body_type vehicle_type  \
0  19UNC1B01HY800062  2017.0  Acura   NSX  Base     Coupe          Car   
1  19UNC1B02HY800023  2017.0  Acura   NSX  Base     Coupe          Car   
2  19UNC1B02HY800071  2017.0  Acura   NSX  Base     Coupe          Car   
3  19UNC1B02LY800001  2020.0  Acura   NSX  Base     Coupe          Car   
4  19UNC1B03HY800063  2017.0  Acura   NSX  Base     Coupe          Car   

  drivetrain transmission  engine_size engine_block  price_range  fuel_M85  \
0        4WD    Automatic          3.5            V            5         0   
1        4WD    Automatic          3.5            V            5         0   
2        4WD    Automatic          3.5            V            5         0   
3        4WD    Automatic          3.5            V            5         0   
4        4WD    Automatic          3.5            V            5         0   

   fuel_Lpg  fuel_Diesel  fuel_Unleaded  fuel_Hydrogen  fuel_PremiumUnleaded  \
0         0            0              0              0                     1   
1         0            0              0              0                     1   
2         0            0              0              0                     1   
3         0            0              0              0                     1   
4         0            0              0              0                     1   

   fuel_Biodiesel  fuel_E85  fuel_Electric  fuel_CompressedNaturalGas  \
0               0         0              1                          0   
1               0         0              1                          0   
2               0         0              1                          0   
3               0         0              1                          0   
4               0         0              1                          0   

   log_miles  
0   9.207035  
1   8.697680  
2  10.095883  
3   8.800566  
4   9.813672

In [342]:
# Sanity Check
joined_df.shape

(2352601, 23)

<div style="border-radius:10px; border:#33a0ff solid; padding: 15px; background-color: #f0f1ff; font-size:100%; text-align:left">
<a class="anchor" id="3-drop-vin">
    <h3> Remove Column - <b><i>vin</i></b> </h3>
</a>
</div>

In [229]:
joined_df.drop(columns=["vin"],inplace = True)


In [343]:
joined_df

vin    year    make           model           trim  \
0    19UNC1B01HY800062  2017.0   Acura             NSX           Base   
1    19UNC1B02HY800023  2017.0   Acura             NSX           Base   
2    19UNC1B02HY800071  2017.0   Acura             NSX           Base   
3    19UNC1B02LY800001  2020.0   Acura             NSX           Base   
4    19UNC1B03HY800063  2017.0   Acura             NSX           Base   
..                 ...     ...     ...             ...            ...   
101  JM3KFBCM6J0334821  2018.0   Mazda            CX-5        Touring   
102  JM3KFADM6J1393032  2018.0   Mazda            CX-5  Grand Touring   
103  JM3KFBBMXJ0337495  2018.0   Mazda            CX-5          Sport   
104  1C4RJEBG6KC731075  2019.0    Jeep  Grand Cherokee        Limited   
105  1NXBR32E65Z404327  2005.0  Toyota         Corolla             LE   

     body_type vehicle_type drivetrain transmission  engine_size engine_block  \
0        Coupe          Car        4WD    Automatic          3.5            V   
1        Coupe          Car        4WD    Automatic          3.5            V   
2        Coupe          Car        4WD    Automatic          3.5            V   
3        Coupe          Car        4WD    Automatic          3.5            V   
4        Coupe          Car        4WD    Automatic          3.5            V   
..         ...          ...        ...          ...          ...          ...   
101        SUV        Truck        4WD    Automatic          2.5            I   
102  Crossover        Truck        FWD    Automatic          2.5            I   
103  Crossover        Truck        4WD    Automatic          2.5            I   
104        SUV        Truck        RWD    Automatic          3.6            V   
105      Sedan          Car        FWD       Manual          1.8            I   

     price_range  fuel_M85  fuel_Lpg  fuel_Diesel  fuel_Unleaded  \
0              5         0         0            0              0   
1              5         0         0            0              0   
2              5         0         0            0              0   
3              5         0         0            0              0   
4              5         0         0            0              0   
..           ...       ...       ...          ...            ...   
101            2         0         0            0              1   
102            2         0         0            0              1   
103            2         0         0            0              1   
104            2         0         0            0              1   
105            1         0         0            0              1   

     fuel_Hydrogen  fuel_PremiumUnleaded  fuel_Biodiesel  fuel_E85  \
0                0                     1               0         0   
1                0                     1               0         0   
2                0                     1               0         0   
3                0                     1               0         0   
4                0                     1               0         0   
..             ...                   ...             ...       ...   
101              0                     0               0         0   
102              0                     0               0         0   
103              0                     0               0         0   
104              0                     0               0         0   
105              0                     0               0         0   

     fuel_Electric  fuel_CompressedNaturalGas  log_miles  
0                1                          0   9.207035  
1                1                          0   8.697680  
2                1                          0  10.095883  
3                1                          0   8.800566  
4                1                          0   9.813672  
..             ...                        ...        ...  
101              0                          0  10.318111  
102              0                     

[Back-to-top](#3-toc)

---

<div style="border-radius:10px; border:#33a0ff solid; padding: 15px; background-color: #f0f1ff; font-size:100%; text-align:left">
<a class="anchor" id="3-split-data">
    <h2> Split Data </h2>
</a>
    
   
<b>Table of Content:</b>
<ul>    
    <li> <a href="#3-split">Split</a></li>
    <li> <a href="#3-san-check-2">Sanity Check</a></li>
    <li> <a href="#3-save">Save</a></li>
</ul>
    
Some of the process later will may induce data leakage. So we split our data into train and test set here. 
    
</div>

<div style="border-radius:10px; border:#33a0ff solid; padding: 15px; background-color: #f0f1ff; font-size:100%; text-align:left">
<a class="anchor" id="3-split">
    <h3> Split</h3>
</a>
</div>

In [344]:
from sklearn.model_selection import train_test_split

X = joined_df.drop(columns=["price_range"])
y = joined_df["price_range"]
X_train, X_test, y_train, y_test = train_test_split(X,y,
                                                    test_size = 0.25,
                                                    stratify=y,
                                                    random_state=42)

# Put together X and y for train data
train_data = X_train
train_data["price_range"] = y_train
# Put together X and y for test data
test_data = X_test
test_data["price_range"] = y_test

<div style="border-radius:10px; border:#33a0ff solid; padding: 15px; background-color: #f0f1ff; font-size:100%; text-align:left">
<a class="anchor" id="3-san-check-2">
    <h3> Sanity Check </h3>
</a>
</div>

In [345]:
train_data.shape

(1764450, 23)

In [346]:
train_data.head()

vin    year    make            model  \
2349744  3TYSZ5AN1LT002321  2020.0  Toyota           Tacoma   
76       1C6RR6LM6JS263153  2018.0     RAM  Ram 1500 Pickup   
89       58ABK1GGXJU087173  2018.0   Lexus               ES   
2814043  5UXKR2C32H0U22684  2017.0     BMW               X5   
90       JTKJF5C74FJ010874  2015.0   Scion               tC   

                       trim body_type vehicle_type drivetrain transmission  \
2349744                 SR5    Pickup        Truck        4WD    Automatic   
76                 Big Horn    Pickup        Truck        RWD    Automatic   
89                      350     Sedan          Car        FWD    Automatic   
2814043           sDrive35i       SUV        Truck        RWD    Automatic   
90       Release Series 9.0     Coupe          Car        FWD    Automatic   

         engine_size engine_block  fuel_M85  fuel_Lpg  fuel_Diesel  \
2349744          3.5            V         0         0            0   
76               3.0            V         0         0            1   
89               3.5            V         0         0            0   
2814043          3.0            I         0         0            0   
90               2.5            I         0         0            0   

         fuel_Unleaded  fuel_Hydrogen  fuel_PremiumUnleaded  fuel_Biodiesel  \
2349744              1              0                     0               0   
76                   0              0                     0               0   
89                   1              0                     0               0   
2814043              0              0                     1               0   
90                   1              0                     0               0   

         fuel_E85  fuel_Electric  fuel_CompressedNaturalGas  log_miles  \
2349744         0              0                          0   8.353497   
76              0              0                          0   9.780189   
89              0              0                          0   9.922555   
2814043         0              0                          0  10.928507   
90              0              0                          0  11.268992   

         price_range  
2349744            3  
76                 3  
89                 3  
2814043            3  
90                 1

In [347]:
test_data.shape

(588151, 23)

In [348]:
test_data.head()

vin    year        make               model  \
2472427  1C4BJWEG2EL188136  2014.0        Jeep  Wrangler Unlimited   
1260954  1HGCS2B84CA006065  2012.0       Honda              Accord   
1714060  3KPF24ADXLE212235  2020.0         Kia               FORTE   
74042    1C4RJFCG2KC813177  2019.0        Jeep      Grand Cherokee   
532197   3VV2B7AX1JM204247  2018.0  Volkswagen              Tiguan   

                  trim body_type vehicle_type drivetrain transmission  \
2472427         Sahara       SUV        Truck        4WD       Manual   
1260954        EX-L V6     Coupe          Car        FWD    Automatic   
1714060            LXS     Sedan          Car        FWD    Automatic   
74042    High Altitude       SUV        Truck        4WD    Automatic   
532197              SE       SUV        Truck        4WD    Automatic   

         engine_size engine_block  fuel_M85  fuel_Lpg  fuel_Diesel  \
2472427          3.6            V         0         0            0   
1260954          3.5            V         0         0            0   
1714060          2.0            I         0         0            0   
74042            3.6            V         0         0            0   
532197           2.0            I         0         0            0   

         fuel_Unleaded  fuel_Hydrogen  fuel_PremiumUnleaded  fuel_Biodiesel  \
2472427              1              0                     0               0   
1260954              1              0                     0               0   
1714060              1              0                     0               0   
74042                1              0                     0               0   
532197               1              0                     0               0   

         fuel_E85  fuel_Electric  fuel_CompressedNaturalGas  log_miles  \
2472427         0              0                          0  11.532738   
1260954         0              0                          0  12.028894   
1714060         0              0                          0  10.063947   
74042           0              0                          0  10.526024   
532197          0              0                          0  10.633834   

         price_range  
2472427            2  
1260954            1  
1714060            2  
74042              4  
532197             2

<div style="border-radius:10px; border:#33a0ff solid; padding: 15px; background-color: #f0f1ff; font-size:100%; text-align:left">
<a class="anchor" id="3-save">
    <h3> Save </h3>
</a>
</div>

In [351]:
# Save test andn train data for modelling
test_data.to_parquet(path = "data/test-data.parquet",index = False)
train_data.to_parquet(path = "data/train-data.parquet",index = False)

[Back-to-top](#3-toc)

---

<div style="border-radius:10px; border: #ffd500 solid; padding: 15px; background-color: #ffffcf; font-size:100%; text-align:left">
<a class="anchor" id="3-learn">
    <h2> Learning / Takeaway </h2>
</a>
</div>

<div style="border-radius:10px; border: #ffd500 solid; padding: 15px; background-color: #ffffcf; font-size:100%; text-align:left">
<a class="anchor">
    <h3> Feature Encoding Methods</h3>
</a>
<br>
    <b>Feature Encoding</b> is a process in machine learning that involves converting categorical or text-based data into numerical format. This is necessary because many machine learning algorithms require numerical input, and encoding allows us to represent categorical information in a way that algorithms can understand. After conducting some research, here is the summary of different imputation methods:
    
<ol>
    <li><b>Label Encoding</b></li>
    <ul>
        <li><p><b>How to do:</b> Label encoding assigns a unique integer to each category in a categorical variable. The integers are usually assigned in ascending order based on the alphabetical order of the categories.</p></li>
        <li><p><b>Pros:</b> Simple and quick to implement. </p></li>
        <li><p><b>Cons:</b> May introduce a <b>false sense</b> of order for nominal variables.</p></li>        
    </ul>
    <li><b>One-hot Encoding</b></li>
    <ul>
        <li><p><b>How to do:</b> One-hot encoding creates binary columns for each category and indicates the presence of the category with a 1 or 0. Each category becomes a new binary feature.</p></li>
        <li><p><b>Pros:</b> Preserves the integrity of the categorical variable.</p></li>
        <li><p><b>Cons:</b> Increases dimensionality for larget dataset. May lead to <b>multicollinearity / dummy variable trap</b>.</p></li>        
    </ul>
    <li><b>Ordinal Encoding (Label Encoding with ordinal meaning)</b></li>
    <ul>
        <li><p><b>How to do:</b>  Ordinal encoding assigns a unique integer to each category based on the specified order or ranking. It is suitable for ordinal variables with a meaningful order.</p></li>
        <li><p><b>Pros:</b> Maintains the order of categories. <b>Reduces dimensionality compared to one-hot encoding.</b></p></li>
        <li><p><b>Cons:</b> Assumes a meaningful order that may not always exist.</p></li>        
    </ul>
    <li><b>Binary Encoding</b></li>
    <ul>
        <li><p><b>How to do:</b> Binary encoding converts each category into binary code. It involves assigning a unique binary code to each category and representing it using a combination of 0s and 1s.</p></li>
        <li><p><b>Pros:</b> Reduces dimensionality compared to one-hot encoding. Captures some ordinal information.</p></li>
        <li><p><b>Cons:</b> Assumes an ordinal relationship that may not always be present.</p></li>        
    </ul>
    <li><b>Frequency Encoding (Count Encoding)</b></li>
    <ul>
        <li><p><b>How to do:</b> Frequency encoding replaces each category with its frequency in the dataset. It is particularly useful for high-cardinality categorical variables.</p></li>
        <li><p><b>Pros:</b> Simplifies high-cardinality variables.</p></li>
        <li><p><b>Cons:</b> May lead to <b>data leakage</b> when using cross validation. May not be suitable for all types of categorical variables.</p></li>        
    </ul>
    <li><b>Target Encoding (Mean Encoding)</b></li>
    <ul>
        <li><p><b>How to do:</b> Target encoding involves replacing each category with the mean of the target variable for that category. It is commonly used for binary classification problems.</p></li>
        <li><p><b>Pros:</b> Encodes information about the relationship between the feature and the target variable.</p></li>
        <li><p><b>Cons:</b> May lead to <b>data leakage</b> when using cross validation. Sensitivity to outliers.</p></li>        
    </ul>
    
</ol>

</div>

[Back-to-top](#3-toc)

---